File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node *, Node *> map;
        
        Node *begin = head;
        while (begin != nullptr) {
            Node *new_node = new Node(begin->val);
            map.insert({begin, new_node});
            begin = begin->next;
        }

        begin = head;
        while (begin != nullptr) {
            Node *curr = map[begin];
            if (begin->next == nullptr) {
                curr->next = nullptr;
            }
            else {
            curr->next = map[begin->next];
            }
            if (begin->random == nullptr) {
                curr->random == nullptr;
            }
            else {
                curr->random = map[begin->random];
            }
            begin = begin->next;
        }
        return map[head];
    }
};

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        if head == None:
            return None
        if head.next == None:
            return head

        while curr != None:
            future = curr.next
            curr.next = prev
            prev = curr
            curr = future
        return prev